ocaml/runtime/shared_heap.c

```diff
@@ -958,8 +1018,417 @@ static void compact_update_ephe_list(volatile value *ephe_p)
+  int free_pools = 0;
+
+  for( int i = 0; i < participating_count; i++ ) {
+    struct caml_heap_state* heap = participants[i]->shared_heap;
+
+    for( int sz = 1; sz < NUM_SIZECLASSES; sz++ ) {
+      /* Just recheck this */
+      CAMLassert(heap->full_pools[sz] == NULL);
+      CAMLassert(heap->avail_pools[sz] == NULL);
+
+      int full_pools_count = compact_count_pools(heap->unswept_full_pools[sz]);
+      int avail_pools_count = compact_count_pools(heap->unswept_avail_pools[sz]);
+
+      pools_count += full_pools_count + avail_pools_count;
+    }
+  }
+
+  /* We need to drain any free_pools that have already been batch allocated. In
+    most cases these will be part of the biggest chunk (as the most recently
+    allocated). */
+  while( pool_freelist.fresh_pools > 0 ) {
+    pool* p = alloc_pool(Caml_state->shared_heap);
+    p->owner = NULL;
+    p->next = pool_freelist.free;
+    pool_freelist.free = p;
+  }
+
+  free_pools = compact_count_pools(pool_freelist.free);
+  pools_count += free_pools;
+
+  pool** all_pools = (pool**)caml_stat_alloc_noexc(sizeof(pool*) * pools_count);
+  int all_pools_idx = 0;
+
+  if( all_pools == NULL ) {
+    caml_gc_log("Unable to allocate all_pools for phase two mark");
+
+    return 0;
+  }
+
+  /* Now we need to go through everything and add all pools to all_pools */
+  for( int i = 0; i < participating_count; i++ ) {
+    struct caml_heap_state* heap = participants[i]->shared_heap;
+
+    for( int sz = 1; sz < NUM_SIZECLASSES; sz++ ) {
+      /* Add all the full pools to all_pools */
+      pool* cur_pool = heap->unswept_full_pools[sz];
+
+      while( cur_pool ) {
+        all_pools[all_pools_idx] = cur_pool;
+        all_pools_idx++;
+        cur_pool = cur_pool->next;
+      }
+
+      /* Add all the avail pools to all_pools */
+      cur_pool = heap->unswept_avail_pools[sz];
+
+      while( cur_pool ) {
+        all_pools[all_pools_idx] = cur_pool;
+        all_pools_idx++;
+        cur_pool = cur_pool->next;
+      }
+    }
+  }
+
+  /* Add all the free pools to all_pools */
+  pool* cur_pool = pool_freelist.free;
+
+  while( cur_pool ) {
+    all_pools[all_pools_idx] = cur_pool;
+    all_pools_idx++;
+    cur_pool = cur_pool->next;
+  }
+
+  CAMLassert(all_pools_idx == pools_count);
+
+  /* Now sort all_pools so that the largest chunks are first */
+  qsort(all_pools, pools_count, sizeof(pool*), compact_compare_pools);
+
+  int live_pools = atomic_load_relaxed(&global_live_pools);
+
+  CAMLassert(live_pools <= pools_count);
+
+  /* We need to mark the first live_pools as live */
+  for(int i = 0; i < live_pools; i++) {
+    pool* p = all_pools[i];
+
+    p->evacuate = 0;
+  }
+
+  /* We now mark everything else as evacuating */
+  for(int i = live_pools; i < pools_count; i++) {
+    pool* p = all_pools[i];
+
+    p->evacuate = 1;
+  }
+
+  /* The last thing we need to do is recreate pool_freelist.free
+     in reverse-sorted order (so we allocate the smallest
+     chunks first) */
```

Comment:
This comment is incorrect! In fact, we recreate `pool_freelist.free` in sorted order, so that we allocate from the _largest_ chunks first). We do this by iterating through the array in reverse order, consing the free list as we go.

Reply:
Fixed